checkResult() should be optional

The `waitForExit()` methods currently internally (all?) use `checkResult()`.

This makes it impossible to wait for a non-sucessful exit - which could be handy for some use cases (e.g. a CLI testing tool).

I think we're running into this issue. I suspect that the following is happening:
```
ManagedProcess process = builder.build();
process.start(); // --> This fails due to some external factor, unknown at the time
process.waitForExit(); // -> This crashes with the error below, due to the cause you mentioned
```

`waitForExit` fails with this:
```
Caused by: ch.vorburger.exec.ManagedProcessException: Asked to waitFor Program [/tmp/75968747865555/MysqlDB4j/base/bin/mysql, -uroot, --socket=/tmp/75968747865555/MysqlDB4j.33803.sock] (in working directory /tmp/75968747865555/MysqlDB4j/base), but it was never even start()'ed! |
  |   | at ch.vorburger.exec.ManagedProcess.waitForExitMaxMsWithoutLog(ManagedProcess.java:439) |  
  |   | at ch.vorburger.exec.ManagedProcess.waitForExit(ManagedProcess.java:418) |  
  |   | at com.airbnb.mysql4j.DB.run(DB.java:405) |  
  |   | at com.airbnb.mysql4j.DB.run(DB.java:374) |  
  |   | at com.airbnb.mysql4j.DB.run(DB.java:357) |  
  |   | at com.airbnb.mysql4j.DB.run(DB.java:365) |  
  |   | at com.airbnb.mysql4j.DB.dropDB(DB.java:418) |  
  |   | at com.airbnb.mysql4j.DB.start(DB.java:181)
```

--- a/src/main/java/ch/vorburger/exec/ExtendedDefaultExecutor.java
+++ b/src/main/java/ch/vorburger/exec/ExtendedDefaultExecutor.java
@@ -1,0 +1,41 @@
+/*
+ * #%L
+ * ch.vorburger.exec
+ * %%
+ * Copyright (C) 2012 - 2023 Michael Vorburger
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package ch.vorburger.exec;
+
+import java.util.function.Function;
+import org.apache.commons.exec.DefaultExecutor;
+
+class ExtendedDefaultExecutor extends DefaultExecutor {
+
+    private Function<Integer, Boolean> exitValueChecker;
+
+    void setIsSuccessExitValueChecker(Function<Integer, Boolean> exitValueChecker) {
+        this.exitValueChecker = exitValueChecker;
+    }
+
+    @Override
+    public boolean isFailure(int exitValue) {
+        if (exitValueChecker == null) {
+            return super.isFailure(exitValue);
+        } else {
+            return !exitValueChecker.apply(exitValue);
+        }
+    }
+}

--- a/src/main/java/ch/vorburger/exec/ManagedProcess.java
+++ b/src/main/java/ch/vorburger/exec/ManagedProcess.java
@@ -32,8 +32,8 @@ import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.function.Function;
 import org.apache.commons.exec.CommandLine;
-import org.apache.commons.exec.DefaultExecutor;
 import org.apache.commons.exec.ExecuteWatchdog;
 import org.apache.commons.exec.Executor;
 import org.apache.commons.exec.ProcessDestroyer;
@@ -67,7 +67,7 @@ public class ManagedProcess implements ManagedProcessState {
     public static final int EXITVALUE_STILL_RUNNING = Executor.INVALID_EXITVALUE - 2;
 
     private final CommandLine commandLine;
-    private final Executor executor = new DefaultExecutor();
+    private final ExtendedDefaultExecutor executor = new ExtendedDefaultExecutor();
     private final StopCheckExecuteWatchdog watchDog = new StopCheckExecuteWatchdog(ExecuteWatchdog.INFINITE_TIMEOUT);
     private final ProcessDestroyer shutdownHookProcessDestroyer = new LoggingShutdownHookProcessDestroyer();
     private final Map<String, String> environment;
@@ -105,7 +105,8 @@ public class ManagedProcess implements ManagedProcessState {
     ManagedProcess(CommandLine commandLine, File directory, Map<String, String> environment,
             InputStream input, boolean destroyOnShutdown, int consoleBufferMaxLines,
             OutputStreamLogDispatcher outputStreamLogDispatcher,
-            List<OutputStream> stdOuts, List<OutputStream> stderr, ManagedProcessListener listener) {
+            List<OutputStream> stdOuts, List<OutputStream> stderr, ManagedProcessListener listener,
+            Function<Integer, Boolean> exitValueChecker) {
         this.commandLine = commandLine;
         this.environment = environment;
         if (input != null) {
@@ -119,6 +120,7 @@ public class ManagedProcess implements ManagedProcessState {
             executor.setWorkingDirectory(directory);
         }
         executor.setWatchdog(watchDog);
+        executor.setIsSuccessExitValueChecker(exitValueChecker);
         this.destroyOnShutdown = destroyOnShutdown;
         this.consoleBufferMaxLines = consoleBufferMaxLines;
         this.outputStreamLogDispatcher = outputStreamLogDispatcher;

--- a/src/main/java/ch/vorburger/exec/ManagedProcessBuilder.java
+++ b/src/main/java/ch/vorburger/exec/ManagedProcessBuilder.java
@@ -26,6 +26,7 @@ import java.io.OutputStream;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
+import java.util.function.Function;
 import org.apache.commons.exec.CommandLine;
 import org.apache.commons.exec.environment.EnvironmentUtils;
 import org.apache.commons.exec.util.StringUtils;
@@ -61,6 +62,7 @@ public class ManagedProcessBuilder {
     protected ManagedProcessListener listener;
     protected List<OutputStream> stdOuts = new ArrayList<>();
     protected List<OutputStream> stdErrs = new ArrayList<>();
+    protected Function<Integer, Boolean> isSuccessExitValueChecker = exitValue -> exitValue == 0;
 
     public ManagedProcessListener getProcessListener() {
         return listener;
@@ -236,7 +238,7 @@ public class ManagedProcessBuilder {
     public ManagedProcess build() {
         return new ManagedProcess(getCommandLine(), directory, environment, inputStream, destroyOnShutdown,
                 consoleBufferMaxLines,
-                outputStreamLogDispatcher, stdOuts, stdErrs, listener);
+                outputStreamLogDispatcher, stdOuts, stdErrs, listener, isSuccessExitValueChecker);
     }
 
     public ManagedProcessBuilder setInputStream(InputStream inputStream) {
@@ -251,6 +253,11 @@ public class ManagedProcessBuilder {
 
     public ManagedProcessBuilder addStdErr(OutputStream stdError) {
         stdErrs.add(stdError);
+        return this;
+    }
+
+    public ManagedProcessBuilder setIsSuccessExitValueChecker(Function<Integer, Boolean> function) {
+        this.isSuccessExitValueChecker = function;
         return this;
     }
 

--- a/src/test/java/ch/vorburger/exec/ManagedProcessTest.java
+++ b/src/test/java/ch/vorburger/exec/ManagedProcessTest.java
@@ -40,28 +40,36 @@ public class ManagedProcessTest {
     @Test
     public void onProcessCompleteInvokedOnCustomListener() throws Exception {
         TestListener listener = new TestListener();
-
         SomeSelfTerminatingExec exec = someSelfTerminatingExec(listener);
         exec.proc.startAndWaitForConsoleMessageMaxMs(exec.msgToWaitFor, 1000);
-        assertNotEquals(Integer.MIN_VALUE, listener.successExitValue);
+        assertNotEquals(Integer.MIN_VALUE, listener.expectedExitValue);
         assertEquals(Integer.MIN_VALUE, listener.failureExitValue);
         assertNull(listener.t);
     }
 
     @Test
-    public void onProcessFailedInvokedOnCustomListener() throws Exception {
+    public void onProcessFailedInvokedOnCustomListenerTraditional() throws Exception {
         TestListener listener = new TestListener();
-
-        SomeSelfTerminatingExec exec = someSelfTerminatingFailingExec(listener);
+        SomeSelfTerminatingExec exec = someSelfTerminatingFailingExec(listener, false);
         try {
             exec.proc.startAndWaitForConsoleMessageMaxMs(exec.msgToWaitFor, 1000);
             fail("Process expected to fail. Should've thrown a ManagedProcessException");
         } catch (ManagedProcessException e) {
-
+            // Expected, ignore
         }
-        assertEquals(Integer.MIN_VALUE, listener.successExitValue);
+        assertEquals(Integer.MIN_VALUE, listener.expectedExitValue);
         assertNotEquals(Integer.MIN_VALUE, listener.failureExitValue);
         assertNotNull(listener.t);
+    }
+
+    @Test
+    public void onProcessFailedInvokedOnCustomListenerWithExitValueChecker() throws Exception {
+        TestListener listener = new TestListener();
+        SomeSelfTerminatingExec exec = someSelfTerminatingFailingExec(listener, true);
+        exec.proc.startAndWaitForConsoleMessageMaxMs(exec.msgToWaitFor, 1000);
+        assertEquals(2, listener.expectedExitValue);
+        assertEquals(Integer.MIN_VALUE, listener.failureExitValue);
+        assertNull(listener.t);
     }
 
     @Test
@@ -200,18 +208,26 @@ public class ManagedProcessTest {
         return r;
     }
 
-    protected SomeSelfTerminatingExec someSelfTerminatingFailingExec(ManagedProcessListener listener)
+    protected SomeSelfTerminatingExec someSelfTerminatingFailingExec(ManagedProcessListener listener,
+            boolean setExitValueChecker)
             throws ManagedProcessException {
+        ManagedProcessBuilder builder;
         SomeSelfTerminatingExec r = new SomeSelfTerminatingExec();
         if (SystemUtils.IS_OS_LINUX || SystemUtils.IS_OS_MAC) {
-            r.proc = new ManagedProcessBuilder("ls").addArgument("-4").setProcessListener(listener).build();
-            r.msgToWaitFor = "usage";
+            builder = new ManagedProcessBuilder("ls").addArgument("-4");
+            // ls (GNU coreutils) 9.1 invoked as "ls -4" prints "ls: invalid option -- '4' \n Try 'ls --help' for more information."
+            r.msgToWaitFor = "invalid option";
         } else if (SystemUtils.IS_OS_WINDOWS) {
-            r.proc = new ManagedProcessBuilder("dir").addArgument("/?").setProcessListener(listener).build();
+            builder = new ManagedProcessBuilder("dir").addArgument("/?");
             r.msgToWaitFor = "Displays a list of files and subdirectories in a directory.";
         } else {
             throw new ManagedProcessException("Unexpected Platform, improve the test dude...");
         }
+
+        if (setExitValueChecker) {
+            builder.setIsSuccessExitValueChecker(exitValue -> exitValue != null);
+        }
+        r.proc = builder.setProcessListener(listener).build();
 
         return r;
     }
@@ -236,13 +252,13 @@ public class ManagedProcessTest {
     }
 
     class TestListener implements ManagedProcessListener {
-        int successExitValue = Integer.MIN_VALUE;
+        int expectedExitValue = Integer.MIN_VALUE;
         int failureExitValue = Integer.MIN_VALUE;
         Throwable t;
 
         @Override
         public void onProcessComplete(int exitValue) {
-            successExitValue = exitValue;
+            expectedExitValue = exitValue;
         }
 
         @Override

